File: java/Samples/Biometrics/VB/SimpleFacesSampleVB/VerifyFace.vb
Imports System
Imports System.Windows.Forms
Imports Neurotec.Biometrics
Imports Neurotec.Biometrics.Client
Imports Neurotec.Biometrics.Gui

Partial Public Class VerifyFace
	Inherits UserControl
#Region "Public constructor"

	Public Sub New()
		InitializeComponent()
	End Sub

#End Region

#Region "Private fields"

	Private _biometricClient As NBiometricClient
	Private _subject1 As NSubject
	Private _subject2 As NSubject

#End Region

#Region "Public properties"

	Public Property BiometricClient() As NBiometricClient
		Get
			Return _biometricClient
		End Get
		Set(ByVal value As NBiometricClient)
			_biometricClient = value
		End Set
	End Property

#End Region

#Region "Private methods"

	Private Function OpenImageTemplate(ByVal faceView As NFaceView, <System.Runtime.InteropServices.Out()> ByRef subject As NSubject) As String
		subject = Nothing
		faceView.Face = Nothing
		msgLabel.Text = String.Empty
		Dim fileLocation As String = String.Empty

		OpenFileDialog.FileName = Nothing
		OpenFileDialog.Title = "Open Template"
		If OpenFileDialog.ShowDialog() = DialogResult.OK Then ' load template
			fileLocation = OpenFileDialog.FileName

			' Check if given file is a template
			Try
				subject = NSubject.FromFile(OpenFileDialog.FileName)
				EnableVerifyButton()
			Catch
			End Try

			' If file is not a template, try to load it as image
			If subject Is Nothing Then
				' Create a face object
				Dim face As New NFace() With {.FileName = fileLocation}
				faceView.Face = face
				subject = New NSubject()
				subject.Faces.Add(face)

				' Extract a template from the subject
				_biometricClient.BeginCreateTemplate(subject, AddressOf OnCreationCompleted, Nothing)
			End If
		End If
		Return fileLocation
	End Function

	Private Sub OnCreationCompleted(ByVal r As IAsyncResult)
		If InvokeRequired Then
			BeginInvoke(New AsyncCallback(AddressOf OnCreationCompleted), r)
		Else
			Try
				Dim status As NBiometricStatus = _biometricClient.EndCreateTemplate(r)
				If status <> NBiometricStatus.Ok Then
					MessageBox.Show(String.Format("The template was not extracted: {0}.", status), Text, MessageBoxButtons.OK, MessageBoxIcon.Error)
				End If
				EnableVerifyButton()
			Catch ex As Exception
				Utils.ShowException(ex)
			End Try
		End If
	End Sub

	Private Sub EnableVerifyButton()
		verifyButton.Enabled = IsSubjectValid(_subject1) AndAlso IsSubjectValid(_subject2)
	End Sub

	Private Function IsSubjectValid(ByVal subject As NSubject) As Boolean
		Return subject IsNot Nothing AndAlso (subject.Status = NBiometricStatus.Ok OrElse subject.Status = NBiometricStatus.MatchNotFound OrElse subject.Status = NBiometricStatus.None AndAlso subject.GetTemplateBuffer() IsNot Nothing)
	End Function

	Private Sub OnVerifyCompleted(ByVal r As IAsyncResult)
		If InvokeRequired Then
			BeginInvoke(New AsyncCallback(AddressOf OnVerifyCompleted), r)
		Else
			Try
				Dim status As NBiometricStatus = _biometricClient.EndVerify(r)
				Dim verificationStatus = String.Format("Verification status: {0}", status)
				If status = NBiometricStatus.Ok Then
					'get matching score
					Dim score As Integer = _subject1.MatchingResults(0).Score
					Dim msg As String = String.Format("Score of matched templates: {0}", score)
					msgLabel.Text = msg
					MessageBox.Show(String.Format("{0}{1}{2}", verificationStatus, Environment.NewLine, msg))
				Else
					MessageBox.Show(verificationStatus)
				End If
			Catch ex As Exception
				Utils.ShowException(ex)
			End Try
		End If
	End Sub

#End Region

#Region "Private form events"

	Private Sub OpenImageButton1Click(ByVal sender As Object, ByVal e As EventArgs) Handles openImageButton1.Click
		templateLeftLabel.Text = String.Empty
		templateLeftLabel.Text = OpenImageTemplate(faceView1, _subject1)
	End Sub

	Private Sub OpenImageButton2Click(ByVal sender As Object, ByVal e As EventArgs) Handles openImageButton2.Click
		templateRightLabel.Text = String.Empty
		templateRightLabel.Text = OpenImageTemplate(faceView2, _subject2)
	End Sub

	Private Sub VerifyFaceLoad(ByVal sender As Object, ByVal e As EventArgs) Handles MyBase.Load
		msgLabel.Text = String.Empty
		templateLeftLabel.Text = String.Empty
		templateRightLabel.Text = String.Empty
		Try
			matchingFarComboBox.BeginUpdate()
			matchingFarComboBox.Items.Add(0.001.ToString("P1"))
			matchingFarComboBox.Items.Add(0.0001.ToString("P2"))
			matchingFarComboBox.Items.Add(0.00001.ToString("P3"))
		Finally
			matchingFarComboBox.EndUpdate()
		End Try

		matchingFarComboBox.SelectedIndex = 1
	End Sub

	Private Sub MatchingFarComboBoxEnter(ByVal sender As Object, ByVal e As EventArgs) Handles matchingFarComboBox.Enter
		defaultButton.Enabled = True
	End Sub

	Private Sub DefaultButtonClick(ByVal sender As Object, ByVal e As EventArgs) Handles defaultButton.Click
		matchingFarComboBox.SelectedIndex = 1
		defaultButton.Enabled = False
	End Sub

	Private Sub ClearImagesButtonClick(ByVal sender As Object, ByVal e As EventArgs) Handles clearImagesButton.Click
		_subject1 = Nothing
		_subject2 = Nothing
		verifyButton.Enabled = False

		faceView1.Face = Nothing
		faceView2.Face = Nothing

		msgLabel.Text = String.Empty
		templateLeftLabel.Text = String.Empty
		templateRightLabel.Text = String.Empty
	End Sub

	Private Sub VerifyButtonClick(ByVal sender As Object, ByVal e As EventArgs) Handles verifyButton.Click
		If _subject1 IsNot Nothing AndAlso _subject2 IsNot Nothing Then
			_biometricClient.BeginVerify(_subject1, _subject2, AddressOf OnVerifyCompleted, Nothing)
		End If
		verifyButton.Enabled = False
	End Sub

	Private Sub MatchingFarComboBoxLeave(ByVal sender As Object, ByVal e As EventArgs) Handles matchingFarComboBox.Leave
		Try
			_biometricClient.MatchingThreshold = Utils.MatchingThresholdFromString(matchingFarComboBox.Text)
			matchingFarComboBox.Text = Utils.MatchingThresholdToString(_biometricClient.MatchingThreshold)
			EnableVerifyButton()
		Catch
			matchingFarComboBox.Select()
			MessageBox.Show("FAR is not valid", Text, MessageBoxButtons.OK, MessageBoxIcon.Error)
		End Try
	End Sub

#End Region
End Class

File: java/Samples/Biometrics/CS/SimpleFacesSampleCS/VerifyFace.cs
using System;
using System.Windows.Forms;
using Neurotec.Biometrics;
using Neurotec.Biometrics.Client;
using Neurotec.Biometrics.Gui;

namespace Neurotec.Samples
{
	public partial class VerifyFace : UserControl
	{
		#region Public constructor

		public VerifyFace()
		{
			InitializeComponent();
		}

		#endregion

		#region Private fields

		private NBiometricClient _biometricClient;
		private NSubject _subject1;
		private NSubject _subject2;

		private int _defaultFar;

		#endregion

		#region Public properties

		public NBiometricClient BiometricClient
		{
			get { return _biometricClient; }
			set
			{
				_biometricClient = value;
				_defaultFar = _biometricClient.MatchingThreshold;
				matchingFarComboBox.Text = Utils.MatchingThresholdToString(_defaultFar);
			}
		}

		#endregion

		#region Private methods

		private string OpenImageTemplate(NFaceView faceView, out NSubject subject)
		{
			subject = null;
			faceView.Face = null;
			msgLabel.Text = string.Empty;
			string fileLocation = string.Empty;

			openFileDialog.FileName = null;
			openFileDialog.Title = @"Open Template";
			if (openFileDialog.ShowDialog() == DialogResult.OK) // load template
			{
				fileLocation = openFileDialog.FileName;

				// Check if given file is a template
				try
				{
					subject = NSubject.FromFile(openFileDialog.FileName);
					EnableVerifyButton();
				}
				catch { }

				// If file is not a template, try to load it as image
				if (subject == null)
				{
					// Create a face object
					NFace face = new NFace() { FileName = fileLocation };
					faceView.Face = face;
					subject = new NSubject();
					subject.Faces.Add(face);

					// Extract a template from the subject
					_biometricClient.BeginCreateTemplate(subject, OnCreationCompleted, null);
				}
			}
			return fileLocation;
		}

		private void OnCreationCompleted(IAsyncResult r)
		{
			if (InvokeRequired)
			{
				BeginInvoke(new AsyncCallback(OnCreationCompleted), r);
			}
			else
			{
				try
				{
					NBiometricStatus status = _biometricClient.EndCreateTemplate(r);
					if (status != NBiometricStatus.Ok)
					{
						MessageBox.Show(string.Format("The template was not extracted: {0}.", status), Text, MessageBoxButtons.OK, MessageBoxIcon.Error);
					}
					EnableVerifyButton();
				}
				catch (Exception ex)
				{
					Utils.ShowException(ex);
				}
			}
		}

		private void EnableVerifyButton()
		{
			verifyButton.Enabled = IsSubjectValid(_subject1) && IsSubjectValid(_subject2);
		}

		private bool IsSubjectValid(NSubject subject)
		{
			return subject != null && (subject.Status == NBiometricStatus.Ok
				|| subject.Status == NBiometricStatus.MatchNotFound
				|| subject.Status == NBiometricStatus.None && subject.GetTemplateBuffer() != null);
		}

		private void OnVerifyCompleted(IAsyncResult r)
		{
			if (InvokeRequired)
			{
				BeginInvoke(new AsyncCallback(OnVerifyCompleted), r);
			}
			else
			{
				try
				{
					NBiometricStatus status = _biometricClient.EndVerify(r);
					var verificationStatus = string.Format("Verification status: {0}", status);
					if (status == NBiometricStatus.Ok)
					{
						//get matching score
						int score = _subject1.MatchingResults[0].Score;
						string msg = string.Format("Score of matched templates: {0}", score);
						msgLabel.Text = msg;
						MessageBox.Show(string.Format("{0}\n{1}", verificationStatus, msg));
					}
					else MessageBox.Show(verificationStatus);
				}
				catch (Exception ex)
				{
					Utils.ShowException(ex);
				}
			}
		}

		#endregion

		#region Private form events

		private void OpenImageButton1Click(object sender, EventArgs e)
		{
			templateLeftLabel.Text = string.Empty;
			templateLeftLabel.Text = OpenImageTemplate(faceView1, out _subject1);
		}

		private void OpenImageButton2Click(object sender, EventArgs e)
		{
			templateRightLabel.Text = string.Empty;
			templateRightLabel.Text = OpenImageTemplate(faceView2, out _subject2);
		}

		private void VerifyFaceLoad(object sender, EventArgs e)
		{
			msgLabel.Text = string.Empty;
			templateLeftLabel.Text = string.Empty;
			templateRightLabel.Text = string.Empty;
			try
			{
				matchingFarComboBox.BeginUpdate();
				matchingFarComboBox.Items.Add(0.001.ToString("P1"));
				matchingFarComboBox.Items.Add(0.0001.ToString("P2"));
				matchingFarComboBox.Items.Add(0.00001.ToString("P3"));
			}
			finally
			{
				matchingFarComboBox.EndUpdate();
			}
		}

		private void MatchingFarComboBoxEnter(object sender, EventArgs e)
		{
			defaultButton.Enabled = true;
		}

		private void DefaultButtonClick(object sender, EventArgs e)
		{
			matchingFarComboBox.Text = Utils.MatchingThresholdToString(_defaultFar);
			defaultButton.Enabled = false;
		}

		private void ClearImagesButtonClick(object sender, EventArgs e)
		{
			_subject1 = null;
			_subject2 = null;
			verifyButton.Enabled = false;

			faceView1.Face = null;
			faceView2.Face = null;

			msgLabel.Text = string.Empty;
			templateLeftLabel.Text = string.Empty;
			templateRightLabel.Text = string.Empty;
		}

		private void VerifyButtonClick(object sender, EventArgs e)
		{
			if (_subject1 != null && _subject2 != null)
			{
				_biometricClient.BeginVerify(_subject1, _subject2, OnVerifyCompleted, null);
			}
			verifyButton.Enabled = false;
		}

		private void MatchingFarComboBoxLeave(object sender, EventArgs e)
		{
			try
			{
				_biometricClient.MatchingThreshold = Utils.MatchingThresholdFromString(matchingFarComboBox.Text);
				matchingFarComboBox.Text = Utils.MatchingThresholdToString(_biometricClient.MatchingThreshold);
				EnableVerifyButton();
			}
			catch
			{
				matchingFarComboBox.Select();
				MessageBox.Show(@"FAR is not valid", Text, MessageBoxButtons.OK, MessageBoxIcon.Error);
			}
		}

		#endregion
	}
}
